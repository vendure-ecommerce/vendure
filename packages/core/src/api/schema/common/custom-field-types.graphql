interface CustomField {
    name: String!
    type: String!
    list: Boolean!
    label: [LocalizedString!]
    description: [LocalizedString!]
    readonly: Boolean
    internal: Boolean
}

type StringCustomFieldConfig implements CustomField {
    name: String!
    type: String!
    list: Boolean!
    length: Int
    label: [LocalizedString!]
    description: [LocalizedString!]
    readonly: Boolean
    internal: Boolean
    pattern: String
    options: [StringFieldOption!]
}

type StringFieldOption {
    value: String!
    label: [LocalizedString!]
}

type LocaleStringCustomFieldConfig implements CustomField {
    name: String!
    type: String!
    list: Boolean!
    length: Int
    label: [LocalizedString!]
    description: [LocalizedString!]
    readonly: Boolean
    internal: Boolean
    pattern: String
}
type IntCustomFieldConfig implements CustomField {
    name: String!
    type: String!
    list: Boolean!
    label: [LocalizedString!]
    description: [LocalizedString!]
    readonly: Boolean
    internal: Boolean
    min: Int
    max: Int
    step: Int
}
type FloatCustomFieldConfig implements CustomField {
    name: String!
    type: String!
    list: Boolean!
    label: [LocalizedString!]
    description: [LocalizedString!]
    readonly: Boolean
    internal: Boolean
    min: Float
    max: Float
    step: Float
}
type BooleanCustomFieldConfig implements CustomField {
    name: String!
    type: String!
    list: Boolean!
    label: [LocalizedString!]
    description: [LocalizedString!]
    readonly: Boolean
    internal: Boolean
}
"""
Expects the same validation formats as the `<input type="datetime-local">` HTML element.
See https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/datetime-local#Additional_attributes
"""
type DateTimeCustomFieldConfig implements CustomField {
    name: String!
    type: String!
    list: Boolean!
    label: [LocalizedString!]
    description: [LocalizedString!]
    readonly: Boolean
    internal: Boolean
    min: String
    max: String
    step: Int
}

type RelationCustomFieldConfig implements CustomField {
    name: String!
    type: String!
    list: Boolean!
    label: [LocalizedString!]
    description: [LocalizedString!]
    readonly: Boolean
    internal: Boolean
    entity: String!
    scalarFields: [String!]!
}

type LocalizedString {
    languageCode: LanguageCode!
    value: String!
}

union CustomFieldConfig =
      StringCustomFieldConfig
    | LocaleStringCustomFieldConfig
    | IntCustomFieldConfig
    | FloatCustomFieldConfig
    | BooleanCustomFieldConfig
    | DateTimeCustomFieldConfig
    | RelationCustomFieldConfig
