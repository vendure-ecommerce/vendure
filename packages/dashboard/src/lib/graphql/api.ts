import type { TypedDocumentNode } from '@graphql-typed-document-node/core';
import { AwesomeGraphQLClient } from 'awesome-graphql-client';
import { DocumentNode, print } from 'graphql';
import { uiConfig } from 'virtual:vendure-ui-config';

const API_URL = uiConfig.apiHost + (uiConfig.apiPort !== 'auto' ? `:${uiConfig.apiPort}` : '') + '/admin-api';

export type Variables = object;
export type RequestDocument = string | DocumentNode;

const awesomeClient = new AwesomeGraphQLClient({
    endpoint: API_URL,
    fetch: async (url: string, options: RequestInit = {}) => {
        // Get the active channel token from localStorage
        const channelToken = localStorage.getItem('vendure-selected-channel-token');
        const headers = new Headers(options.headers);

        if (channelToken) {
            headers.set('vendure-token', channelToken);
        }

        return fetch(url, {
            ...options,
            headers,
            credentials: 'include',
            mode: 'cors',
        });
    },
});

export type VariablesAndRequestHeadersArgs<V extends Variables> =
    V extends Record<any, never>
        ? [variables?: V, requestHeaders?: HeadersInit]
        : [variables: V, requestHeaders?: HeadersInit];

function query<T, V extends Variables = Variables>(
    document: RequestDocument | TypedDocumentNode<T, V>,
    variables?: V,
): Promise<T> {
    const documentString = typeof document === 'string' ? document : print(document);
    return awesomeClient.request(documentString, variables) as any;
}

function mutate<T, V extends Variables = Variables>(
    document: TypedDocumentNode<T, V>,
): (variables: V) => Promise<T>;
function mutate(document: RequestDocument): (variables: Variables) => Promise<unknown>;
function mutate<T, V extends Variables = Variables>(
    document: TypedDocumentNode<T, V>,
    variables: V,
): Promise<T>;
function mutate(document: RequestDocument, variables: Variables): Promise<unknown>;
function mutate<T, V extends Variables = Variables>(
    document: RequestDocument | TypedDocumentNode<T, V>,
    maybeVariables?: V,
): Promise<T> | ((variables: V) => Promise<T>) {
    const documentString = typeof document === 'string' ? document : print(document);
    if (maybeVariables) {
        return awesomeClient.request(documentString, maybeVariables) as any;
    } else {
        return (variables: V): Promise<T> => {
            return awesomeClient.request(documentString, variables) as any;
        };
    }
}

export const api = {
    query,
    mutate,
};
