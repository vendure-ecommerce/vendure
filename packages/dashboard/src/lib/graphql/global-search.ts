import { gql } from 'graphql-tag';

// Note: These GraphQL definitions are placeholders for the frontend implementation.
// The actual backend GraphQL schema will be implemented separately according to
// the GLOBAL_SEARCH_IMPLEMENTATION_PLAN.md

export const GLOBAL_SEARCH_QUERY = gql`
    query GlobalSearch($input: GlobalSearchInput!) {
        globalSearch(input: $input) {
            id
            type
            title
            subtitle
            description
            url
            thumbnailUrl
            metadata
            relevanceScore
            lastModified
        }
    }
`;

// TypeScript types that match the expected GraphQL schema
export interface GlobalSearchInput {
    query: string;
    types?: string[];
    limit?: number;
    skip?: number;
    channelId?: string;
}

export interface GlobalSearchResult {
    id: string;
    type: string;
    title: string;
    subtitle?: string;
    description?: string;
    url: string;
    thumbnailUrl?: string;
    metadata?: Record<string, any>;
    relevanceScore?: number;
    lastModified?: string;
}
