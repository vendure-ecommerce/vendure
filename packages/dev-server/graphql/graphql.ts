import type { introspection } from './graphql-env.d.ts';
import { initGraphQLTada } from 'gql.tada';

export const graphql = initGraphQLTada<{
    disableMasking: true;
    introspection: introspection;
    scalars: {
        DateTime: string;
        JSON: any;
        Money: number;
    };
}>();

export type { FragmentOf, ResultOf, VariablesOf } from 'gql.tada';
export { readFragment } from 'gql.tada';
