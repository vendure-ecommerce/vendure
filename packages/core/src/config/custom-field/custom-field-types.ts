import {
    BooleanCustomFieldConfig as GraphQLBooleanCustomFieldConfig,
    CustomField,
    DateTimeCustomFieldConfig as GraphQLDateTimeCustomFieldConfig,
    FloatCustomFieldConfig as GraphQLFloatCustomFieldConfig,
    IntCustomFieldConfig as GraphQLIntCustomFieldConfig,
    LocaleStringCustomFieldConfig as GraphQLLocaleStringCustomFieldConfig,
    LocalizedString,
    RelationCustomFieldConfig as GraphQLRelationCustomFieldConfig,
    StringCustomFieldConfig as GraphQLStringCustomFieldConfig,
} from '@vendure/common/lib/generated-types';
import { CustomFieldsObject, CustomFieldType, Type } from '@vendure/common/lib/shared-types';

import { Injector } from '../../common/injector';
import { VendureEntity } from '../../entity/base/base.entity';

// prettier-ignore
export type DefaultValueType<T extends CustomFieldType> =
    T extends 'string' | 'localeString' ? string :
        T extends 'int' | 'float' ? number :
            T extends 'boolean' ? boolean :
                T extends 'datetime' ? Date : never;

export type BaseTypedCustomFieldConfig<T extends CustomFieldType, C extends CustomField> = Omit<
    C,
    '__typename' | 'list'
> & {
    type: T;
    /**
     * @description
     * Whether or not the custom field is available via the Shop API.
     * @default true
     */
    public?: boolean;
    nullable?: boolean;
};

/**
 * @description
 * Configures a custom field on an entity in the {@link CustomFields} config object.
 *
 * @docsCategory custom-fields
 */
export type TypedCustomSingleFieldConfig<
    T extends CustomFieldType,
    C extends CustomField
> = BaseTypedCustomFieldConfig<T, C> & {
    list?: false;
    defaultValue?: DefaultValueType<T>;
    validate?: (
        value: DefaultValueType<T>,
        injector: Injector,
    ) => string | LocalizedString[] | void | Promise<string | LocalizedString[] | void>;
};

export type TypedCustomListFieldConfig<
    T extends CustomFieldType,
    C extends CustomField
> = BaseTypedCustomFieldConfig<T, C> & {
    list?: true;
    defaultValue?: Array<DefaultValueType<T>>;
    validate?: (value: Array<DefaultValueType<T>>) => string | LocalizedString[] | void;
};

export type TypedCustomFieldConfig<
    T extends CustomFieldType,
    C extends CustomField
> = BaseTypedCustomFieldConfig<T, C> &
    (TypedCustomSingleFieldConfig<T, C> | TypedCustomListFieldConfig<T, C>);

export type StringCustomFieldConfig = TypedCustomFieldConfig<'string', GraphQLStringCustomFieldConfig>;
export type LocaleStringCustomFieldConfig = TypedCustomFieldConfig<
    'localeString',
    GraphQLLocaleStringCustomFieldConfig
>;
export type IntCustomFieldConfig = TypedCustomFieldConfig<'int', GraphQLIntCustomFieldConfig>;
export type FloatCustomFieldConfig = TypedCustomFieldConfig<'float', GraphQLFloatCustomFieldConfig>;
export type BooleanCustomFieldConfig = TypedCustomFieldConfig<'boolean', GraphQLBooleanCustomFieldConfig>;
export type DateTimeCustomFieldConfig = TypedCustomFieldConfig<'datetime', GraphQLDateTimeCustomFieldConfig>;
export type RelationCustomFieldConfig = TypedCustomFieldConfig<
    'relation',
    Omit<GraphQLRelationCustomFieldConfig, 'entity' | 'scalarFields'>
> & { entity: Type<VendureEntity>; graphQLType?: string; eager?: boolean };

/**
 * @description
 * An object used to configure a custom field.
 *
 * @docsCategory custom-fields
 */
export type CustomFieldConfig =
    | StringCustomFieldConfig
    | LocaleStringCustomFieldConfig
    | IntCustomFieldConfig
    | FloatCustomFieldConfig
    | BooleanCustomFieldConfig
    | DateTimeCustomFieldConfig
    | RelationCustomFieldConfig;

/**
 * @description
 * Most entities can have additional fields added to them by defining an array of {@link CustomFieldConfig}
 * objects on against the corresponding key.
 *
 * ### Configuration options
 *
 * All custom fields share some common properties:
 *
 * * `name: string`: The name of the field
 * * `type: string`: A string of type {@link CustomFieldType}
 * * `list: boolean`: If set to `true`, then the field will be an array of the specified type
 * * `label?: LocalizedString[]`: An array of localized labels for the field.
 * * `description?: LocalizedString[]`: An array of localized descriptions for the field.
 * * `public?: boolean`: Whether or not the custom field is available via the Shop API. Defaults to `true`
 * * `readonly?: boolean`: Whether or not the custom field can be updated via the GraphQL APIs. Defaults to `false`
 * * `internal?: boolean`: Whether or not the custom field is exposed at all via the GraphQL APIs. Defaults to `false`
 * * `defaultValue?: any`: The default value when an Entity is created with this field.
 * * `nullable?: boolean`: Whether the field is nullable in the database. If set to `false`, then a `defaultValue` should be provided.
 * * `validate?: (value: any) => string | LocalizedString[] | void`: A custom validation function. If the value is valid, then
 *     the function should not return a value. If a string or LocalizedString array is returned, this is interpreted as an error message.
 *
 * The `LocalizedString` type looks like this:
 *
 * ```TypeScript
 * type LocalizedString = {
 *   languageCode: LanguageCode;
 *   value: string;
 * };
 * ```
 *
 * In addition to the common properties, the following field types have some type-specific properties:
 *
 * #### `string` type
 *
 * * `pattern?: string`: A regex pattern which the field value must match
 * * `options?: { value: string; label?: LocalizedString[]; };`: An array of pre-defined options for the field.
 * * `length?: number`: The max length of the varchar created in the database. Defaults to 255. Maximum is 65,535.
 *
 * #### `localeString` type
 *
 * * `pattern?: string`: A regex pattern which the field value must match
 * * `length?: number`: The max length of the varchar created in the database. Defaults to 255. Maximum is 65,535.
 *
 * #### `int` & `float` type
 *
 * * `min?: number`: The minimum permitted value
 * * `max?: number`: The maximum permitted value
 * * `step?: number`: The step value
 *
 * #### `datetime` type
 *
 * The min, max & step properties for datetime fields are intended to be used as described in
 * [the datetime-local docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/datetime-local#Additional_attributes)
 *
 * * `min?: string`: The earliest permitted date
 * * `max?: string`: The latest permitted date
 * * `step?: string`: The step value
 *
 * #### `relation` type
 *
 * * `entity: VendureEntity`: The entity which this custom field is referencing
 * * `eager?: boolean`: Whether to [eagerly load](https://typeorm.io/#/eager-and-lazy-relations) the relation. Defaults to false.
 * * `graphQLType?: string`: The name of the GraphQL type that corresponds to the entity.
 *     Can be omitted if it is the same, which is usually the case.
 *
 * @example
 * ```TypeScript
 * bootstrap({
 *     // ...
 *     customFields: {
 *         Product: [
 *             { name: 'infoUrl', type: 'string' },
 *             { name: 'downloadable', type: 'boolean', defaultValue: false },
 *             { name: 'shortName', type: 'localeString' },
 *         ],
 *         User: [
 *             { name: 'socialLoginToken', type: 'string', public: false },
 *         ],
 *     },
 * })
 * ```
 *
 * @docsCategory custom-fields
 */
export interface CustomFields {
    Address?: CustomFieldConfig[];
    Administrator?: CustomFieldConfig[];
    Asset?: CustomFieldConfig[];
    Channel?: CustomFieldConfig[];
    Collection?: CustomFieldConfig[];
    Customer?: CustomFieldConfig[];
    Facet?: CustomFieldConfig[];
    FacetValue?: CustomFieldConfig[];
    Fulfillment?: CustomFieldConfig[];
    GlobalSettings?: CustomFieldConfig[];
    Order?: CustomFieldConfig[];
    OrderLine?: CustomFieldConfig[];
    Product?: CustomFieldConfig[];
    ProductOption?: CustomFieldConfig[];
    ProductOptionGroup?: CustomFieldConfig[];
    ProductVariant?: CustomFieldConfig[];
    User?: CustomFieldConfig[];
    ShippingMethod?: CustomFieldConfig[];
}

/**
 * This interface should be implemented by any entity which can be extended
 * with custom fields.
 */
export interface HasCustomFields {
    customFields: CustomFieldsObject;
}
