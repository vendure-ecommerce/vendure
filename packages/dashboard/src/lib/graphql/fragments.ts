import { graphql, ResultOf } from './graphql.js';

export const assetFragment = graphql(`
    fragment Asset on Asset {
        id
        createdAt
        updatedAt
        name
        fileSize
        mimeType
        type
        preview
        source
        width
        height
        focalPoint {
            x
            y
        }
    }
`);

export type AssetFragment = ResultOf<typeof assetFragment>;

export const configurableOperationFragment = graphql(`
    fragment ConfigurableOperation on ConfigurableOperation {
        args {
            name
            value
        }
        code
    }
`);

export type ConfigurableOperationFragment = ResultOf<typeof configurableOperationFragment>;

export const configArgDefinitionFragment = graphql(`
    fragment ConfigArgDefinition on ConfigArgDefinition {
        name
        type
        required
        defaultValue
        list
        ui
        label
        description
    }
`);

export const configurableOperationDefFragment = graphql(
    `
        fragment ConfigurableOperationDef on ConfigurableOperationDefinition {
            args {
                ...ConfigArgDefinition
            }
            code
            description
        }
    `,
    [configArgDefinitionFragment],
);

export const errorResultFragment = graphql(`
    fragment ErrorResult on ErrorResult {
        errorCode
        message
    }
`);

export type ConfigurableOperationDefFragment = ResultOf<typeof configurableOperationDefFragment>;
export type ConfigArgDefFragment = ResultOf<typeof configArgDefinitionFragment>;
