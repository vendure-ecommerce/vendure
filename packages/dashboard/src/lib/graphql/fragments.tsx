import { graphql, ResultOf } from "./graphql.js";

export const assetFragment = graphql(`
    fragment Asset on Asset {
        id
        createdAt
        updatedAt
        name
        fileSize
        mimeType
        type
        preview
        source
        width
        height
        focalPoint {
            x
            y
        }
    }
`);

export type AssetFragment = ResultOf<typeof assetFragment>;

export const configurableOperationFragment = graphql(`
    fragment ConfigurableOperation on ConfigurableOperation {
        args {
            name
            value
        }
        code
    }
`);

export type ConfigurableOperationFragment = ResultOf<typeof configurableOperationFragment>;

export const configurableOperationDefFragment = graphql(`
    fragment ConfigurableOperationDef on ConfigurableOperationDefinition {
        args {
            name
            type
            required
            defaultValue
            list
            ui
            label
            description
        }
        code
        description
    }
`);

export const errorResultFragment = graphql(`
    fragment ErrorResult on ErrorResult {
        errorCode
        message
    }
`);


export type ConfigurableOperationDefFragment = ResultOf<typeof configurableOperationDefFragment>;
