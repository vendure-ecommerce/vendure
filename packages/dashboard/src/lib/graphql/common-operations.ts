import { configArgDefinitionFragment } from '@/vdb/graphql/fragments.js';

import { graphql } from './graphql.js';

export const duplicateEntityDocument = graphql(`
    mutation DuplicateEntity($input: DuplicateEntityInput!) {
        duplicateEntity(input: $input) {
            ... on DuplicateEntitySuccess {
                newEntityId
            }
            ... on ErrorResult {
                errorCode
                message
            }
            ... on DuplicateEntityError {
                duplicationError
            }
        }
    }
`);

export const getEntityDuplicatorsDocument = graphql(
    `
        query GetEntityDuplicators {
            entityDuplicators {
                code
                description
                requiresPermission
                forEntities
                args {
                    ...ConfigArgDefinition
                }
            }
        }
    `,
    [configArgDefinitionFragment],
);
