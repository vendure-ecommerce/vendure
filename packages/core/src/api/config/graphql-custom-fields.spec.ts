import { printSchema } from 'graphql';
import { describe, expect, it } from 'vitest';

import { CustomFieldConfig, CustomFields } from '../../config/custom-field/custom-field-types';

import {
    addGraphQLCustomFields,
    addOrderLineCustomFieldsInput,
    addRegisterCustomerCustomFieldsInput,
} from './graphql-custom-fields';

describe('addGraphQLCustomFields()', () => {
    it('uses JSON scalar if no custom fields defined', () => {
        const input = `
            type Product {
                id: ID
            }
        `;
        const customFieldConfig: CustomFields = {
            Product: [],
        };
        const result = addGraphQLCustomFields(input, customFieldConfig, false);
        expect(printSchema(result)).toMatchSnapshot();
    });

    it('extends a type', () => {
        const input = `
            type Product {
                id: ID
            }
        `;
        const customFieldConfig: CustomFields = {
            Product: [{ name: 'available', type: 'boolean' }],
        };
        const result = addGraphQLCustomFields(input, customFieldConfig, false);
        expect(printSchema(result)).toMatchSnapshot();
    });

    it('extends a type with a translation', () => {
        const input = `
                    type Product {
                        id: ID
                        translations: [ProductTranslation!]!
                    }

                    type ProductTranslation {
                        id: ID
                    }
                `;
        const customFieldConfig: CustomFields = {
            Product: [
                { name: 'available', type: 'boolean' },
                { name: 'shortName', type: 'localeString' },
            ],
        };
        const result = addGraphQLCustomFields(input, customFieldConfig, false);
        expect(printSchema(result)).toMatchSnapshot();
    });

    it('extends a type with a Create input', () => {
        const input = `
                    type Product {
                        id: ID
                    }

                    input CreateProductInput {
                        image: String
                    }
                `;
        const customFieldConfig: CustomFields = {
            Product: [
                { name: 'available', type: 'boolean' },
                { name: 'shortName', type: 'localeString' },
            ],
        };
        const result = addGraphQLCustomFields(input, customFieldConfig, false);
        expect(printSchema(result)).toMatchSnapshot();
    });

    it('extends a type with an Update input', () => {
        const input = `
                    type Product {
                        id: ID
                    }

                    input UpdateProductInput {
                        image: String
                    }
                `;
        const customFieldConfig: CustomFields = {
            Product: [
                { name: 'available', type: 'boolean' },
                { name: 'shortName', type: 'localeString' },
            ],
        };
        const result = addGraphQLCustomFields(input, customFieldConfig, false);
        expect(printSchema(result)).toMatchSnapshot();
    });

    it('extends a type with a Create input and a translation', () => {
        const input = `
                    type Product {
                        id: ID
                    }

                    type ProductTranslation {
                        id: ID
                    }

                    input ProductTranslationInput {
                        id: ID
                    }

                    input CreateProductInput {
                        image: String
                    }
                `;
        const customFieldConfig: CustomFields = {
            Product: [
                { name: 'available', type: 'boolean' },
                { name: 'shortName', type: 'localeString' },
            ],
        };
        const result = addGraphQLCustomFields(input, customFieldConfig, false);
        expect(printSchema(result)).toMatchSnapshot();
    });

    it('extends a type with SortParameters', () => {
        const input = `
                    type Product {
                        id: ID
                    }

                    input ProductSortParameter {
                        id: SortOrder
                    }

                    enum SortOrder {
                        ASC
                        DESC
                    }
                `;
        const customFieldConfig: CustomFields = {
            Product: [
                { name: 'available', type: 'boolean' },
                { name: 'shortName', type: 'localeString' },
            ],
        };
        const result = addGraphQLCustomFields(input, customFieldConfig, false);
        expect(printSchema(result)).toMatchSnapshot();
    });

    it('extends a type with FilterParameters', () => {
        const input = `
                    type Product {
                        name: String
                    }

                    input ProductFilterParameter {
                        id: StringOperators
                    }

                    input StringOperators {
                        eq: String
                    }

                    input NumberOperators {
                        eq: Float
                    }

                    input DateOperators {
                        eq: String
                    }

                    input BooleanOperators {
                        eq: Boolean
                    }

                `;
        const customFieldConfig: CustomFields = {
            Product: [
                { name: 'available', type: 'boolean' },
                { name: 'shortName', type: 'localeString' },
                { name: 'rating', type: 'float' },
                { name: 'published', type: 'datetime' },
            ],
        };
        const result = addGraphQLCustomFields(input, customFieldConfig, false);
        expect(printSchema(result)).toMatchSnapshot();
    });

    it('publicOnly = true', () => {
        const input = `
                 type Product {
                     id: ID
                 }
            `;
        const customFieldConfig: CustomFields = {
            Product: [
                { name: 'available', type: 'boolean', public: true },
                { name: 'profitMargin', type: 'float', public: false },
            ],
        };
        const result = addGraphQLCustomFields(input, customFieldConfig, true);
        expect(printSchema(result)).toMatchSnapshot();
    });

    it('extends OrderAddress if Address custom fields defined', () => {
        const input = `
             type Address {
                 id: ID
                 streetLine1: String
             }

             type OrderAddress {
                 streetLine1: String
             }
        `;
        const customFieldConfig: CustomFields = {
            Address: [{ name: 'instructions', type: 'string' }],
        };
        const result = addGraphQLCustomFields(input, customFieldConfig, true);
        expect(printSchema(result)).toMatchSnapshot();
    });
});

describe('addOrderLineCustomFieldsInput()', () => {
    it('Modifies the schema when the addItemToOrder & adjustOrderLine mutation is present', () => {
        const input = `
            type Mutation {
                addItemToOrder(id: ID!, quantity: Int!): Boolean
                adjustOrderLine(id: ID!, quantity: Int): Boolean
            }
        `;
        const customFieldConfig: CustomFieldConfig[] = [
            { name: 'giftWrap', type: 'boolean' },
            { name: 'message', type: 'string' },
        ];
        const result = addOrderLineCustomFieldsInput(input, customFieldConfig);
        expect(printSchema(result)).toMatchSnapshot();
    });
});

describe('addRegisterCustomerCustomFieldsInput()', () => {
    it('add public writable custom fields to RegisterCustomerInput', () => {
        const input = `
            input RegisterCustomerInput {
                emailAddress: String!
                title: String
                firstName: String
                lastName: String
                phoneNumber: String
                password: String
            }

            type Mutation {
                registerCustomerAccount(input: RegisterCustomerInput!): Boolean!
            }
        `;
        const customFieldConfig: CustomFieldConfig[] = [
            { name: 'isB2B', type: 'boolean', nullable: false },
            { name: 'message', type: 'string' },
            { name: 'rating', type: 'int', public: false },
            { name: 'dbRef', type: 'int', internal: true },
        ];
        const result = addRegisterCustomerCustomFieldsInput(input, customFieldConfig);
        expect(printSchema(result)).toMatchSnapshot();
    });
});
