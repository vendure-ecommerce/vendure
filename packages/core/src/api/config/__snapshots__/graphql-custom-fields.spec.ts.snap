// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`addGraphQLCustomFields() > extends OrderAddress if Address custom fields defined 1`] = `
"type Address {
  id: ID
  streetLine1: String
  customFields: AddressCustomFields
}

type OrderAddress {
  streetLine1: String
  customFields: AddressCustomFields
}

scalar JSON

scalar DateTime

type AddressCustomFields {
  instructions: String
}"
`;

exports[`addGraphQLCustomFields() > extends a type 1`] = `
"type Product {
  id: ID
  customFields: ProductCustomFields
}

scalar JSON

scalar DateTime

type ProductCustomFields {
  available: Boolean
}"
`;

exports[`addGraphQLCustomFields() > extends a type with FilterParameters 1`] = `
"type Product {
  name: String
  customFields: ProductCustomFields
}

input ProductFilterParameter {
  id: StringOperators
  available: BooleanOperators
  shortName: StringOperators
  rating: NumberOperators
  published: DateOperators
}

input StringOperators {
  eq: String
}

input NumberOperators {
  eq: Float
}

input DateOperators {
  eq: String
}

input BooleanOperators {
  eq: Boolean
}

scalar JSON

scalar DateTime

type ProductCustomFields {
  available: Boolean
  shortName: String
  rating: Float
  published: DateTime
}"
`;

exports[`addGraphQLCustomFields() > extends a type with SortParameters 1`] = `
"type Product {
  id: ID
  customFields: ProductCustomFields
}

input ProductSortParameter {
  id: SortOrder
  available: SortOrder
  shortName: SortOrder
}

enum SortOrder {
  ASC
  DESC
}

scalar JSON

scalar DateTime

type ProductCustomFields {
  available: Boolean
  shortName: String
}"
`;

exports[`addGraphQLCustomFields() > extends a type with a Create input 1`] = `
"type Product {
  id: ID
  customFields: ProductCustomFields
}

input CreateProductInput {
  image: String
  customFields: CreateProductCustomFieldsInput
}

scalar JSON

scalar DateTime

type ProductCustomFields {
  available: Boolean
  shortName: String
}

input CreateProductCustomFieldsInput {
  available: Boolean
}"
`;

exports[`addGraphQLCustomFields() > extends a type with a Create input and a translation 1`] = `
"type Product {
  id: ID
  customFields: ProductCustomFields
}

type ProductTranslation {
  id: ID
  customFields: ProductTranslationCustomFields
}

input ProductTranslationInput {
  id: ID
  customFields: ProductTranslationInputCustomFields
}

input CreateProductInput {
  image: String
  customFields: CreateProductCustomFieldsInput
}

scalar JSON

scalar DateTime

type ProductCustomFields {
  available: Boolean
  shortName: String
}

type ProductTranslationCustomFields {
  shortName: String
}

input CreateProductCustomFieldsInput {
  available: Boolean
}

input ProductTranslationInputCustomFields {
  shortName: String
}"
`;

exports[`addGraphQLCustomFields() > extends a type with a translation 1`] = `
"type Product {
  id: ID
  translations: [ProductTranslation!]!
  customFields: ProductCustomFields
}

type ProductTranslation {
  id: ID
  customFields: ProductTranslationCustomFields
}

scalar JSON

scalar DateTime

type ProductCustomFields {
  available: Boolean
  shortName: String
}

type ProductTranslationCustomFields {
  shortName: String
}"
`;

exports[`addGraphQLCustomFields() > extends a type with an Update input 1`] = `
"type Product {
  id: ID
  customFields: ProductCustomFields
}

input UpdateProductInput {
  image: String
  customFields: UpdateProductCustomFieldsInput
}

scalar JSON

scalar DateTime

type ProductCustomFields {
  available: Boolean
  shortName: String
}

input UpdateProductCustomFieldsInput {
  available: Boolean
}"
`;

exports[`addGraphQLCustomFields() > publicOnly = true 1`] = `
"type Product {
  id: ID
  customFields: ProductCustomFields
}

scalar JSON

scalar DateTime

type ProductCustomFields {
  available: Boolean
}"
`;

exports[`addGraphQLCustomFields() > uses JSON scalar if no custom fields defined 1`] = `
"type Product {
  id: ID
  customFields: JSON
}

scalar JSON

scalar DateTime"
`;

exports[`addOrderLineCustomFieldsInput() > Modifies the schema when the addItemToOrder & adjustOrderLine mutation is present 1`] = `
"type Mutation {
  addItemToOrder(id: ID!, quantity: Int!, customFields: OrderLineCustomFieldsInput = null): Boolean
  adjustOrderLine(id: ID!, quantity: Int, customFields: OrderLineCustomFieldsInput = null): Boolean
}

input OrderLineCustomFieldsInput {
  giftWrap: Boolean
  message: String
}"
`;

exports[`addRegisterCustomerCustomFieldsInput() > add public writable custom fields to RegisterCustomerInput 1`] = `
"input RegisterCustomerInput {
  emailAddress: String!
  title: String
  firstName: String
  lastName: String
  phoneNumber: String
  password: String
  customFields: RegisterCustomerCustomFieldsInput
}

type Mutation {
  registerCustomerAccount(input: RegisterCustomerInput!): Boolean!
}

input RegisterCustomerCustomFieldsInput {
  isB2B: Boolean
  message: String
}"
`;
